app/org/sagebionetworks/bridge/dynamodb/DynamoTaskDao.java

```diff
@@ -107,15 +114,21 @@ public void updateTasks(String healthCode, List<Task> tasks) {
                     if (task.getFinishedOn() != null) {
                         dbTask.setFinishedOn(task.getFinishedOn());
                         dbTask.setHidesOn(task.getFinishedOn());
+                        isFinishing = true;
+                    }
+                    // Not transactional, so we need to do these one-by-one, not in the batch API which
+                    // returns incomprehensibly low-level errors. This is what can happen:
+                    // both succeed, both fail, np
+                    // task saved, event fails, no further scheduled tasks. So don't allow that
+                    // event saved, task fails, 2nd future task created, existing task still needs to be saved 
+                    //          This is acceptable, so do this and save event first 
+                    if (isFinishing) {
+                        taskEventService.publishTaskFinishedEvent(dbTask);
```

Comment:
It seems really odd to me that the DynamoTaskDao has a dependency on TaskEventService. The Dynamo DAOs are supposed to encapsulate only the DDB-specific logic for CRUDing DDB objects. I feel like the call to TaskEventService, along with a lot of the other logic for filtering startsOn and finishesOn, could be trivially moved to TaskService.